unittests/util/MockPrometheus.h

```diff
@@ -0,0 +1,181 @@
+            metric = std::make_unique<MetricType>(name, labelsString, impl);
+        } else {
+            auto& impl = counterDoubleImpls[key];
+            metric = std::make_unique<MetricType>(name, labelsString, impl);
+        }
+        auto* ptr = metrics.emplace(key, std::move(metric)).first->second.get();
+        auto metricPtr = dynamic_cast<MetricType*>(ptr);
+        if (metricPtr == nullptr)
+            throw std::runtime_error("Wrong metric type");
+        return *metricPtr;
+    }
+
+    std::unordered_map<std::string, std::unique_ptr<MetricBase>> metrics;
+    std::unordered_map<std::string, ::testing::StrictMock<MockCounterImplInt>> counterIntImpls;
+    std::unordered_map<std::string, ::testing::StrictMock<MockCounterImplUint>> counterUintImpls;
+    std::unordered_map<std::string, ::testing::StrictMock<MockCounterImplDouble>> counterDoubleImpls;
+};
+
+/**
+ * @note this class should be the first in the inheritance list
+ */
+struct WithMockPrometheus : virtual ::testing::Test {
+    WithMockPrometheus()
+    {
+        PrometheusSingleton::replaceInstance(std::make_unique<MockPrometheusImpl>());
+    }
+
+    ~WithMockPrometheus() override
+    {
+        if (HasFailure()) {
+            std::cerr << "Registered metrics:\n";
+            for (auto const& [key, metric] : mockPrometheus().metrics) {
+                std::cout << key << "\n";
+            }
+            std::cerr << "\n";
+        }
+        PROMETHEUS_INIT();
+    }
+
+    static MockPrometheusImpl&
+    mockPrometheus()
+    {
+        auto* ptr = dynamic_cast<MockPrometheusImpl*>(&PROMETHEUS());
+        if (ptr == nullptr)
+            throw std::runtime_error("Wrong prometheus type");
+        return *ptr;
+    }
+
+    template <typename MetricType>
+    static auto&
+    makeMock(std::string name, std::string labelsString)
+    {
+        auto* mockPrometheusPtr = dynamic_cast<MockPrometheusImpl*>(&PROMETHEUS());
+        if (mockPrometheusPtr == nullptr)
+            throw std::runtime_error("Wrong prometheus type");
+
+        std::string const key = name + labelsString;
+        mockPrometheusPtr->makeMetric<MetricType>(std::move(name), std::move(labelsString));
+        if constexpr (std::is_same_v<typename MetricType::ValueType, std::int64_t>) {
+            return mockPrometheusPtr->counterIntImpls[key];
+        } else if constexpr (std::is_same_v<typename MetricType::ValueType, std::uint64_t>) {
+            return mockPrometheusPtr->counterUintImpls[key];
+        } else if constexpr (std::is_same_v<typename MetricType::ValueType, double>) {
+            return mockPrometheusPtr->counterDoubleImpls[key];
+        }
+        throw std::runtime_error("Wrong metric type");
+    }
+};
+
+/**
+ * @note this class should be the first in the inheritance list
+ */
+struct WithPrometheus : virtual ::testing::Test {
+    WithPrometheus()
+    {
+        PROMETHEUS_INIT();
+    }
+
+    ~WithPrometheus() override
+    {
+        PROMETHEUS_INIT();
```

Comment:
Curious why PROMETHEUS_INIT is called in destructor 🤔

Reply:
Because all expectations are being checked in mock destructors. Because counters are owned by static prometheus object they will live until the end of tests or until the next test reinitialising prometheus.  So without reinitialising prometheus in destructor expectations from current test could be checked in some other test.